### ISSUE TITLE
Change Request: `no-useless-assignment` should check only references created by Identifier

### ISSUE THREAD

**Issue by yeonjuan** [eslint/eslint#19160]
### ESLint version

v9.x

### What problem do you want to solve?

I propose that ESLint rules that use scope-manager only handle `Reference`s generated by `Identifier`.

ESLint does not assume any semantics for jsx nodes (reference, scope..).
This is not a problem when using ESLint's scope manager, but when using the custom scope manager in a custom parser, the ESLint core rules that rely on `Reference` don't work well. see https://github.com/typescript-eslint/typescript-eslint/issues/10219#issuecomment-2492086613

It is possible that `Reference`s are generated by the custom parser with a variety of node types, so I propose that ESLint core only handles references generated from Identifiers in the rules.

### What do you think is the correct solution?

https://github.com/eslint/eslint/blob/main/lib/rules/no-useless-assignment.js#L311-L313

```js
const readReferences = targetAssignment.variable.references.filter(reference => reference.isRead() && reference.identifier.type ==="Identifier'');
```

Or, it may be better to wrap the scope-manager used by the core rules so that it only handles references whose node type is Identifier.

### Participation

- [x] I am willing to submit a pull request for this change.

**Comment by nzakas** [comment]
Your description is quite abstract, so I'm not sure I understand what you're asking for or what the problem is. Can you be more specific?

**Comment by bradzacher** [comment]
The full context is in the issue that they linked, but summarising it here:

typescript-eslint's scope manager understands JSX and so it creates `Reference`s for `JSXIdentifier` nodes when appropriate, for example:
```ts
function Component() {}
<Component />; // this creates a reference from the JSXIdentifier to the function's Identifier

function test() {}
<test />; // this does not create a reference, as per JSX spec for lower-case names
```
[typescript-eslint playground with the scope tab open](https://typescript-eslint.io/play/#ts=5.5.2&showAST=scope&fileType=.tsx&code=LAKAZgrgdgxgLgSwPZQAQGEkFsAOKCmUcAFAJSoDeqAvqADya4FGoD0AfANyiiSyIpUcfAGcS5KrRB1hYtl1BA&eslintrc=N4KABGBEBOCuA2BTAzpAXGYBfEWg&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)

By-and-large this works just fine and all rules function as expected.

However `no-useless-assignment` makes an assumption that `References` will only contain `Identifiers` thanks to this selector:
https://github.com/eslint/eslint/blob/576bcc5461c0c00c30dfceec9abcddb99e559c74/lib/rules/no-useless-assignment.js#L457-L466

When this assumption is broken it causes [a false positive from the `no-useless-assignments` lint rule](https://typescript-eslint.io/play/#ts=5.6.2&fileType=.tsx&code=KYDwDg9gTgLgBAE2AMwIYFcA29noHYDGMAlhHnAIJhgAUAlHAN4CwAUHHAWQM7wDCZGKmJ5gUAOrEYACwDKQkgQDyUAHIQAClAhhucALxwARAmIA3IwG42HKMBjoo5GjY5wAPAD5XbjwLxCImKSMvKoiirqWjrc3uy%2BboxGYYpwMKAwRgC%2BPm7uAPT%2BgaISUnIKxMpqmtq6cb4F9XB01qw5rEA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQHYHsBaWZRJZZQgQwoEsBzXAW0VwBd0pFpp9pJwYAL4ghQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false).

---

The proposal is just to more explicitly encode the `Identifier` assumption into the rule as shown above.
This would allow the rule to function correctly "for free" and would avoid us needing to create an entire extension rule to handle this case

**Comment by bradzacher** [comment]
That's a good point @mdjermanovic - I'm not sure what the best behaviour would be.

Right now the issue is that the rule sees the reference but it doesn't setup the state - meaning that you have this disconnect between the two causing the false positive.

If we just go with filtering the non-identifier references then at least the variable would be treated akin to an unused variable and would not report unless it was used in a non-JSX way. I honestly would say that would probably be good enough of a fix to cover the gap. It would leave the little gap you mentioned but that's getting into 1-in-a-trillion usecase so probably okay to ignore.

> Would it maybe be better to check if the variable has at least one non-identifier reference, and in that case ignore the variable because the rule can't reliably work with it?

If you want to opt-out entirely this could be a good option! I'm okay with either :)

**Comment by mdjermanovic** [comment]
> Is there any effect on `/* global */` and `/* exported */` variables?

`/* exported */` variables are already ignored by this rule because marking variables as such means they're also used elsewhere, so we generally can't know if those uses occur in runtime between the uses we can see in the file being linted.

Similar for `/* global *`/ variables - since they're declared and possibly used elsewhere, this rule ignores them.

**Comment by mdjermanovic** [comment]
> Right now the issue is that the rule sees the reference but it doesn't setup the state - meaning that you have this disconnect between the two causing the false positive.
> 
> If we just go with filtering the non-identifier references then at least the variable would be treated akin to an unused variable and would not report unless it was used in a non-JSX way. I honestly would say that would probably be good enough of a fix to cover the gap. It would leave the little gap you mentioned but that's getting into 1-in-a-trillion usecase so probably okay to ignore.
> 
> > Would it maybe be better to check if the variable has at least one non-identifier reference, and in that case ignore the variable because the rule can't reliably work with it?
> 
> If you want to opt-out entirely this could be a good option! I'm okay with either :)

Given that this rule already ignores variables marked by `markVariableAsUsed()` (for example, by the `react/jsx-uses-vars` rule), I think it should just ignore variables with any non-identifier references as well because in both cases the usage analysis isn't reliable.

**Comment by mdjermanovic** [comment]
I updated the title to clarify that this issue is specifically about the `no-useless-assignment` rule.

@nzakas what do you think about updating this rule to just entirely ignore variables that have at least one non-identifier reference?
